### ISSUE TITLE
Allow Minetest to be installed to a readonly location and write to user dir

### ISSUE THREAD

**Issue by neuhaus** [minetest/minetest#4528]
Fresh install of minetest 0.4.14 64bit (**using the installer**) on Windows 10 64bit Pro.

I get this crash message:

> 2016-09-18 11:03:20: ERROR[Main]: An unhandled exception occurred: Failed to open log file debug.txt: Permission denied
> 2016-09-18 11:03:20: ERROR[Main]: In thread 13fc:
> 2016-09-18 11:03:20: ERROR[Main]: /home/stefan/mt-build/build/minetest_64/minetest/src/main.cpp:f6: main: A fatal error occured: Failed to open log file debug.txt: Permission denied
> 2016-09-18 11:03:20: ERROR[Main]: Debug stacks:
> 2016-09-18 11:03:20: ERROR[Main]: DEBUG STACK FOR THREAD 13fc:
> 2016-09-18 11:03:20: ERROR[Main]: #0  main
> 
> This application has requested the Runtime to terminate it in an unusual way.
> Please contact the application's support team for more information.

**Comment by neuhaus** [comment]
If I navigate to a writable directory in CMD and then start the exe directly it launches.
Then, when I try to generate a new world I get the next fatal error:

> 2016-09-18 11:15:53: ERROR[Main]: Angegebener Weltpfad existiert nicht: c:\Program Files\Minetest\bin\..\worlds\world
> 2016-09-18 11:15:53: ERROR[Main]: Error writing configuration file: "c:\Program Files\Minetest\bin\..\minetest.conf"

The first error messages translated to english:
"The given world path does not exist: c:\Program Files\Minetest\bin\..\worlds\world"

**Comment by sfan5** [comment]
Looks like some path handling code broke

**Comment by paramat** [comment]
Blocker?

**Comment by neuhaus** [comment]
Looks like a blocker to me

**Comment by sfan5** [comment]
@paramat depends on whether this bug is avoidable by the installer (e.g. by installing to somewhere else)

**Comment by neuhaus** [comment]
I installed it to the default location. If that means it doesn't work.. well that is a blocker in my book.

**Comment by Fixer-007** [comment]
I think it is a blocker, people just install in default DIR, if it does not work in default DIR -> broken installer.

**Comment by garywhite207** [comment]
I do know that it does not do this with the portable 64bit version in Windows 7 Professional, so this appears to be a Windows 10 only issue.

**Comment by Zeno-** [comment]
I can fix the debug.txt issue but I suspect that if debug.txt cannot be written to then any other file that needs to be written to will have permission problems as well, so I think this should be classified as an installer issue

**Comment by kaeza** [comment]
So after some [discussion](http://irc.minetest.net/minetest-dev/2016-12-19#i_4766801), there seem to be several places to blame.

The proposed solution for the upcoming release is to modify the installer script so it installs to -- and sets shortcuts to run from -- `%APPDATA%\Minetest` (so it is in a writable location), and provide a proper fix to this for 0.4.16.

cc: @red-001, @rubenwardy

**Comment by rubenwardy** [comment]
The official installer defaults to c://games/Minetest not c://Program Files/Minetest

**Comment by rubenwardy** [comment]
I do agree however with writing to app data rather than failing to write to program data. C://games is more of a workaround

**Comment by neuhaus** [comment]
Good. The installer should use `%PROGRAMFILES%`.  I'm sure it's in a Microsoft Application Guide somewhere, a quick web search did not find it.

Anyway, i don't think people want to create a extra folder in the root directory of their bootdrive when they already have a folder for programs.

**Comment by neuhaus** [comment]
@garywhite207 I think you are not having this issue on Windows 7 only because you use the portable version. If you used the normal version and install it in the `%PROGRAMFILES%` directory you would see the same issue on Windows 7 as well (and Windows 8 etc).

**Comment by Zeno-** [comment]
At this point in time there are two ways to fix this:

a) Don't provide an official installer; or
b) Make the installer behave as discussed

The third option, which is to make all paths compatible with an "installer version/build" of the program, would delay the release for a considerable amount of time (and should, IMO, be deferred to the next release)

**Comment by MillersMan** [comment]
Not sure but I would expect %APPDATA% to be problematic on multi-user systems. E.g. shortcuts for all users that only work for a specific one, etc. Based on my experience from writing other installers but can't tell to what degree it affects Minetest.

**Comment by garywhite207** [comment]
The issue is that the C:\Program Files and Program Files (x86) folders on Windows 7/8/10 require administrator permission to write to the file. I installed the 64-bit Minetest 0.4.14 onto Windows 7 Professional 64 bit, and tried to run, and I got the same message that @neuhaus received. However, when I right-clicked and clicked "Run as administrator", it was able to write to the debug file and launch the client. So the issue is not with the installer, it's that Minetest needs administrator privileges to run if installed to C:\Program Files or C:\Program Files (x86).

**Comment by kaeza** [comment]
@garywhite207 As discussed on IRC, there are two related issues:

* Minetest writes things to `.` (the current working directory), so it fails if launched from a read-only directory. Minetest on Windows should be the same as in Linux, with system files (binaries, builtin, etc.) in `C:\Program Files[ x86]` (`/usr` on Linux), and the user files (mods, worlds, etc.) in `%APPDATA%` (`~/.minetest`).
* The installer sets the shortcut's CWD to `C:\Program Files`. This should not be an issue. See my comment above.

@MillersMan Installing binaries to `%APPDATA%` is not the ideal solution, but given that Minetest 0.4.15 is about to be released, changing things in the main codebase is not an option, so this is a workaround. A proper fix will be done for 0.4.16.

**Comment by garywhite207** [comment]
Please note: If this happens writing to /bin/debug.txt, then that means all the folders (/mods, /games, /worlds) are completely inoperable because the User Account Control (Windows Vista and newer) would prevent it unless you give administrator permission. This may need until 0.4.16 to correct the issue, cause this is a Windows issue that just altering the installer will not fix.

**Comment by MillersMan** [comment]
kaeza: A speedhack could be to write a preloader that unzips or copies everything to %APPDATA% and starts from there. That tool could be installed at any location. Did something like that to construct an msi installer based on the system configuration before starting it.

**Comment by paramat** [comment]
We'll work on this after release.

**Comment by rubenwardy** [comment]
duplicate of #4931 (other issue has better discussion)
